tools/cleanup/cleanup.go

```diff
@@ -0,0 +1,308 @@
+	}
+	gcrrepo, err := name.NewRepository(repoRoot)
+	if err != nil {
+		return 0, errors.Wrapf(err, "cannot open registry %q", repoRoot)
+	}
+	count := 0
+	// Walk down the registry, checking all images and deleting the old ones.
+	return count, google.Walk(gcrrepo, func(repo name.Repository, tags *google.Tags, err error) error {
+		// If we got an error, just return it, there's nothing to do here.
+		if tags == nil {
+			return err
+		}
+		for k, m := range tags.Manifests {
+			ref := repo.String() + "@" + k
+			age := int(time.Since(m.Uploaded).Hours() / 24)
+			log.Printf("%q is %d days old", ref, age)
+			if m.Uploaded.Before(before) {
+				if *dryRun {
+					log.Printf("[DRY RUN] delete %q", ref)
+				} else {
+					// Delete all tags first, otherwise the image can't be deleted.
+					for _, tag := range m.Tags {
+						if err := deleteImage(repo.String() + ":" + tag); err != nil {
+							return err
+						}
+					}
+					if err := deleteImage(ref); err != nil {
+						return err
+					}
+					log.Printf("Deleted %q (uploaded on %s)", ref, m.Uploaded)
+					count++
+				}
+			}
+		}
+		return nil
+	}, google.WithAuthFromKeychain(auther))
+}
+
+// deleteResources call resourceDeleter functions in parallel, one for each given project.
+func deleteResources(projects []string, before time.Time, deleteFunc resourceDeleter) (int, []string) {
+	// Map of errors returned from functions, to avoid repeated entries.
+	var errs = make(map[string]error)
+
+	// Locks for the concurrent tasks.
+	var errsLock sync.RWMutex
+	var countLock sync.RWMutex
+	var wg sync.WaitGroup
+
+	// Blocking channel to keep concurrency under control.
+	concurrencyChan := make(chan struct{}, *concurrentOperations)
+	defer close(concurrencyChan)
+
+	count := 0
+	for i := range projects {
+		wg.Add(1)
+		go func(project string) {
+			defer wg.Done()
+			// Block processing if concurrency reached the limit.
+			concurrencyChan <- struct{}{}
+			// Do not process if previous invocations failed. This prevents a large
+			// build-up of failed requests and rate limit exceeding (e.g. bad auth).
+			errsLock.RLock()
+			failed := len(errs) > 0
+			errsLock.RUnlock()
+			if failed {
+				return
+			}
+			c, err := deleteFunc(project, before)
+			// Update counter and errors list.
+			countLock.Lock()
+			count += c
+			countLock.Unlock()
+			if err != nil {
+				cause := errors.Cause(err).Error()
+				errsLock.Lock()
+				if _, ok := errs[cause]; !ok {
+					errs[cause] = err
+				}
+				errsLock.Unlock()
+			}
+			<-concurrencyChan
+		}(projects[i])
+	}
+	wg.Wait()
+	// Extract the error strings from the map of errors.
+	errStrings := make([]string, 0)
+	for i := range errs {
+		errStrings = append(errStrings, errs[i].Error())
+	}
+	return count, errStrings
+}
+
+// showStats displays basic stats based on the return values from deleteResources.
+func showStats(count int, errs []string) {
+	log.Printf("%d resources deleted", count)
```

Comment:
not sure if count is needed if it's just for displaying how many resources deleted, does it worth throttling with another lock with just this information?

Reply:
It's not needed, but it's a nice to have statistics. Deleting a resource is several orders of magnitude slower than a simple locking, the throttling you're concerned about is negligible.